spec/lib/nerve/reporter_zookeeper_spec.rb

```diff
@@ -29,5 +32,109 @@
+      @reporter.report_up
+    end
+
+    after(:each) do
+      # reset the class variable to avoid mock object zk leak
+      Nerve::Reporter::Zookeeper.class_variable_set(:@@zk_pool, {})
+      Nerve::Reporter::Zookeeper.class_variable_set(:@@zk_pool_count, {})
+    end
+
+    it "returns true on report_up" do
+      expect(@reporter.report_up).to be true
+    end
+
+    it "returns true on report_down" do
+      expect(@reporter.report_down).to be true
+    end
+
+    it "returns true on ping?" do
+      expect(zk).to receive(:exists?) { "zk_path" }.and_return(true)
+      expect(@reporter.ping?).to be true
+    end
+
+    context "when zk.connected? started to return false" do
+      it 'returns false on ping?' do
+        # this condition is triggered if connection has been lost for a while (a few sec)
+        expect(zk).to receive(:connected?).and_return(false)
+        expect(zk).not_to receive(:exists?)
+        expect(@reporter.ping?).to be false
+      end
+
+      it 'returns false on report_up without zk operation' do
+        # this condition is triggered if connection has been lost for a while (a few sec)
+        expect(zk).to receive(:connected?).and_return(false)
+        expect(zk).not_to receive(:set)
+        expect(@reporter.report_up).to be false
+      end
+
+      it 'returns false on report_up without zk operation' do
+        # this condition is triggered if connection has been lost for a while (a few sec)
+        expect(zk).to receive(:connected?).and_return(false)
+        expect(zk).not_to receive(:delete)
+        expect(@reporter.report_down).to be false
+      end
+    end
+
+    context "when there is a short disconnection" do
+      it 'returns false on ping?' do
+        # this condition is triggered if connection is shortly interrupted
+        # so connected? still return true
+        expect(zk).to receive(:exists?).and_raise(ZK::Exceptions::OperationTimeOut)
+        expect(@reporter.ping?).to be false
+      end
+
+      it 'swallows zk connetion errors and returns false on report_up' do
+        # this condition is triggered if connection is shortly interrupted
+        # so connected? still return true
+        expect(zk).to receive(:set).and_raise(ZK::Exceptions::OperationTimeOut)
+        expect(@reporter.report_up).to be false
+      end
+
+      it 'swallows zk connetion errors and returns false on report_down' do
+        # this condition is triggered if connection is shortly interrupted
+        # so connected? still return true
+        expect(zk).to receive(:delete).and_raise(ZK::Exceptions::OperationTimeOut)
+        expect(@reporter.report_down).to be false
+      end
+    end
+
+    context "when there is other ZK errors" do
+      it 'raises zk non-connection error on ping?' do
+        # this condition is triggered if connection is shortly interrupted
+        # so connected? still return true
+        expect(zk).to receive(:exists?).and_raise(ZK::Exceptions::SessionExpired)
```

Comment:
here too

Reply:
for report_up and report_down, they actually throw exception from a different zk operation, not zk.exists?. so there is no need to have a common before block